amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java

```diff
@@ -0,0 +1,451 @@
+                        entry.getDataBuffer(),
+                        Optional.empty(), null, Schema.BYTES, 0, true, -1L);
+                message.getMessageBuilder().clearSequenceId();
+                props = message.getMessageBuilder().getPropertiesList().stream()
+                        .collect(Collectors.toMap(KeyValue::getKey, KeyValue::getValue));
+            } catch (Exception e) {
+                log.error("Deserialize entry dataBuffer failed for exchange {}, skip it first.",
+                        exchange.getName(), e);
+                PENDING_SIZE_UPDATER.decrementAndGet(this);
+                entry.release();
+                continue;
+            }
+            Set<Destination> destinations = getDestinations(
+                    props.getOrDefault(MessageConvertUtils.PROP_ROUTING_KEY, ""), getMessageHeaders());
+
+            final Position position = entry.getPosition();
+
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            if (!destinations.isEmpty()) {
+                if (destinations.size() > 1) {
+                    entry.getDataBuffer().retain(destinations.size() - 1);
+                }
+                final int readerIndex = message.getDataBuffer().readerIndex();
+                for (Destination des : destinations) {
+                    futures.add(sendMessage(message, des, readerIndex));
+                }
+            }
+            entry.release();
+            FutureUtil.waitForAll(futures).whenComplete((__, t) -> {
+                if (t != null) {
+                    log.error("Failed to route message {} for exchange {}.", position, exchange.exchangeName, t);
+                    cursor.rewind();
+                    tryToReadMoreEntries();
+                    return;
+                }
+
+                cursor.asyncDelete(position, new AsyncCallbacks.DeleteCallback() {
+                    @Override
+                    public void deleteComplete(Object ctx) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("{} Deleted message at {}", exchange.getName(), ctx);
+                        }
+                    }
+
+                    @Override
+                    public void deleteFailed(ManagedLedgerException exception, Object ctx) {
+                        log.error("{} Failed to delete message at {}", exchange.getName(), ctx, exception);
+                    }
+                }, entry.getPosition());
+                tryToReadMoreEntries();
+            });
+        }
+    }
+
+    private void tryToReadMoreEntries() {
+        if (PENDING_SIZE_UPDATER.decrementAndGet(this) < replicatorQueueSize * 0.5
+                && HAVE_PENDING_READ_UPDATER.get(this) == FALSE) {
+            this.readMoreEntries();
+        }
+    }
+
+    private CompletableFuture<Void> sendMessage(MessageImpl<byte[]> msg, Destination des, int readerIndex) {
+        return getProducer(des.name, des.type)
```

Comment:
ByteBuf has retained(), we should release ByteBuf if we failed to create the producer or send the message.

Reply:
I change to use sync method to create producer, please take a look again.